File: sdk/media-sdk.js
export class MediaClient {
  constructor(baseURL) {
    this.baseURL = baseURL;
    this.token = null; // (1h)
  }

  async login(username, password) {
    const res = await fetch(`${this.baseURL}/auth/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username, password }),
    });

    if (!res.ok) {
      throw new Error("Login falhou: " + res.status);
    }

    const data = await res.json();
    this.token = data.token;
    return data.token;
  }

  async sign(path) {
    if (!this.token) throw new Error("Faça login primeiro");

    const res = await fetch(`${this.baseURL}/media/sign`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.token}`,
      },
      body: JSON.stringify({ path }),
    });

    if (!res.ok) {
      throw new Error("Erro ao gerar signed URL: " + res.status);
    }

    const data = await res.json();
    return data.signed_url;
  }

  async getBlob(path) {
    const signedUrl = await this.sign(path);
    const res = await fetch(signedUrl);

    if (!res.ok) {
      throw new Error("Erro ao baixar: " + res.status);
    }

    return await res.blob();
  }

  async getStream(path) {
    const signedUrl = await this.sign(path);

    const res = await fetch(signedUrl);
    if (!res.ok) {
      throw new Error("Erro ao baixar stream: " + res.status);
    }

    return res.body;
  }

  async getSignedUrl(path) {
    return this.sign(path);
  }
}

File: sdk/media_sdk.py
import aiohttp
import asyncio
from typing import AsyncGenerator

class MediaClient:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")
        self.token = None
        self.session = aiohttp.ClientSession()

    # -------------------------
    # LOGIN
    # -------------------------
    async def login(self, username: str, password: str):
        url = f"{self.base_url}/auth/login"
        async with self.session.post(url, json={
            "username": username,
            "password": password
        }) as res:
            if res.status != 200:
                raise Exception(f"Login falhou: {res.status}")
            data = await res.json()
            self.token = data["token"]
            return self.token

    # -------------------------
    # SIGN URL
    # -------------------------
    async def sign(self, path: str) -> str:
        if not self.token:
            raise Exception("Faça login primeiro")

        url = f"{self.base_url}/media/sign"
        async with self.session.post(
            url,
            json={"path": path},
            headers={"Authorization": f"Bearer {self.token}"}
        ) as res:
            if res.status != 200:
                raise Exception(f"Erro ao assinar: {res.status}")
            data = await res.json()
            return data["signed_url"]

    # -------------------------
    # GET BYTES
    # -------------------------
    async def get_bytes(self, path: str) -> bytes:
        signed = await self.sign(path)
        async with self.session.get(signed) as res:
            if res.status != 200:
                raise Exception(f"Erro ao baixar mídia: {res.status}")
            return await res.read()

    # -------------------------
    # STREAM (ideal p/ vídeos)
    # -------------------------
    async def stream(self, path: str) -> AsyncGenerator[bytes, None]:
        signed = await self.sign(path)
        async with self.session.get(signed) as res:
            if res.status != 200:
                raise Exception(f"Erro no stream: {res.status}")
            async for chunk in res.content.iter_chunked(4096):
                yield chunk

    # -------------------------
    # SALVAR ARQUIVO
    # -------------------------
    async def save(self, path: str, dest: str):
        data = await self.get_bytes(path)
        with open(dest, "wb") as f:
            f.write(data)
        return dest

    async def close(self):
        await self.session.close()
